### PR TITLE
Implement bech32m

### PR DESCRIPTION
**Pull request by benthecarman**
Implements bech32 in accordance to https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/Address.scala
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/Address.scala:231]: Doesn't this have to be a `WitnessScriptPubKeyV1`? 

I guess we don't have that now, but my understanding is v0 will always use the old bech32 encoding?

Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/Address.scala:231]: ah i see below, we should probably add a TODO or file an issue

File: core/src/main/scala/org/bitcoins/core/util/Bech32.scala
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/util/Bech32.scala:14]: add a link to what/where these differences are specified. I'm assuming they are on in the BIP?

File: core-test/src/test/scala/org/bitcoins/core/protocol/Bech32mSpec.scala
Comment (Christewart) [inline core-test/src/test/scala/org/bitcoins/core/protocol/Bech32mSpec.scala:11]: nit: This is the old style of test cases (uses scalacheck, not scalatest). We are trying to slowly move away from these over the long term. Can you write these using `BitcoinSUnitTest` ? Also convert the `Bech32Spec` file if you are feeling ambitious :-)

Comment (Christewart) [reply core-test/src/test/scala/org/bitcoins/core/protocol/Bech32mSpec.scala:11]: https://github.com/bitcoin-s/bitcoin-s/blob/c5be81d5e21cd22a4e69dc86dcf47fbaa7d623f3/core-test/src/test/scala/org/bitcoins/core/protocol/AddressTest.scala#L12

Comment (Christewart) [reply core-test/src/test/scala/org/bitcoins/core/protocol/Bech32mSpec.scala:11]: I guess this means this file can be deleted and test cases migrated to `Bech32mTest`

**Review by Christewart — CHANGES_REQUESTED** [review]
So what are your thoughts on merging this? Should we wait until we have `WitnessScriptPubKeyV1` available to us ?

If i understand this, that would mean doing Tapscript/Taproot integration?

**Review by Christewart — APPROVED** [review]
So this LGTM, 

but is BIP350 actually finalized? Is it deployed as part of bitcoin core 0.21.0? It seems to me this shouldn't be merged until the spec is finalized ? 

https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki

**Comment by benthecarman** [comment]
> So this LGTM,
> 
> but is BIP350 actually finalized? Is it deployed as part of bitcoin core 0.21.0? It seems to me this shouldn't be merged until the spec is finalized ?
> 
> [bitcoin/bips@`master`/bip-0350.mediawiki](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)

The spec should be finalized, it is not merged into core yet: https://github.com/bitcoin/bitcoin/pull/20861

imo this should be fine to merge though

**Comment by Christewart** [comment]
> > So this LGTM,
> > but is BIP350 actually finalized? Is it deployed as part of bitcoin core 0.21.0? It seems to me this shouldn't be merged until the spec is finalized ?
> > [bitcoin/bips@`master`/bip-0350.mediawiki](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)
> 
> The spec should be finalized, it is not merged into core yet: [bitcoin/bitcoin#20861](https://github.com/bitcoin/bitcoin/pull/20861)
> 
> imo this should be fine to merge though

I vote to wait until it gets merged into core as there is less likely to be changes after that occurs

**Pull request merged**
